temporal-serviceclient/src/main/java/io/temporal/serviceclient/SystemInfoInterceptor.java

```diff
@@ -0,0 +1,104 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.serviceclient;
+
+import io.grpc.*;
+import io.temporal.api.workflowservice.v1.GetSystemInfoRequest;
+import io.temporal.api.workflowservice.v1.GetSystemInfoResponse;
+import io.temporal.api.workflowservice.v1.WorkflowServiceGrpc;
+import java.util.concurrent.CompletableFuture;
+import javax.annotation.Nullable;
+
+public class SystemInfoInterceptor implements ClientInterceptor {
+
+  private final CompletableFuture<GetSystemInfoResponse.Capabilities> serverCapabilitiesFuture;
+
+  public SystemInfoInterceptor(
+      CompletableFuture<GetSystemInfoResponse.Capabilities> serverCapabilitiesFuture) {
+    this.serverCapabilitiesFuture = serverCapabilitiesFuture;
+  }
+
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+      MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+    return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(
+        next.newCall(method, callOptions)) {
+
+      @Override
+      public void start(Listener<RespT> responseListener, Metadata headers) {
+        if (!serverCapabilitiesFuture.isDone()) {
+          if (method == WorkflowServiceGrpc.getGetSystemInfoMethod()) {
+            // It's already a getSystemsInfo call, so we just listen on it and populate the
+            // capabilities
+            responseListener =
+                new ForwardingClientCallListener.SimpleForwardingClientCallListener<RespT>(
+                    responseListener) {
+                  @Override
+                  public void onMessage(RespT message) {
+                    if (message instanceof GetSystemInfoResponse) {
+                      GetSystemInfoResponse response = (GetSystemInfoResponse) message;
+                      serverCapabilitiesFuture.complete(response.getCapabilities());
+                    }
+                    super.onMessage(message);
+                  }
+
+                  @Override
+                  public void onClose(Status status, Metadata trailers) {
+                    if (Status.UNIMPLEMENTED.getCode().equals(status.getCode())) {
+                      serverCapabilitiesFuture.complete(
+                          GetSystemInfoResponse.Capabilities.getDefaultInstance());
+                    }
+                    super.onClose(status, trailers);
+                  }
+                };
+          } else {
+            // Need to reach system capabilities, so make a getSystemInfo call in a blocking manner.
+            // We don't try to squash into one and optimize the several getSystemInfo calls that may
```

Comment:
> We don't try to squash into one and optimize the several getSystemInfo calls

:+1: Same as Go. If in a racy situation the this code is reached by multiple calls and we have to make multiple getSystemInfo calls and overwrite that future one right after another, that's much more preferable than locking. https://github.com/temporalio/sdk-go/blob/942edfb5dce0dbff49e19f002470726660d54b53/internal/internal_workflow_client.go#L901-L904

Reply:
Yeah, in Java the first who completes wins. Great, thanks for confirming that it's aligned!